File: src/indexer/adapter/php.ts
import Parser from "tree-sitter";
import type { Tree } from "tree-sitter";
import type { LanguageAdapter } from "./LanguageAdapter.js";
import {
  getParser,
  clearCache as clearGrammarCache,
  createQuery,
} from "../treesitter/grammarLoader.js";
import type {
  ExtractedSymbol,
  ExtractedCall,
} from "../treesitter/extractCalls.js";
import type { ExtractedImport } from "../treesitter/extractImports.js";
import { logger } from "../../util/logger.js";
import { findEnclosingSymbol as findEnclosingSymbolUtil } from "../treesitter/symbolUtils.js";

class PhpAdapter implements LanguageAdapter {
  languageId = "php";
  fileExtensions = [".php", ".phtml"] as const;

  private parser: Parser | null = null;

  getParser(): Parser | null {
    if (!this.parser) {
      this.parser = getParser("php");
    }
    return this.parser;
  }

  parse(content: string, _filePath: string): Tree | null {
    const parser = this.getParser();
    if (!parser) {
      return null;
    }

    try {
      // Use 1MB buffer to handle files >32KB (tree-sitter default limit)
      const tree = parser.parse(content, undefined, {
        bufferSize: 1024 * 1024,
      });

      if (!tree) {
        return null;
      }

      if (tree.rootNode.hasError) {
        logger.warn(
          "Syntax errors detected in PHP file - attempting partial extraction",
          { filePath: _filePath },
        );
      }

      return tree;
    } catch (error) {
      logger.error("Failed to parse PHP file", {
        filePath: _filePath,
        error,
      });
      return null;
    }
  }

  extractSymbols(
    tree: Tree,
    _content: string,
    filePath: string,
  ): ExtractedSymbol[] {
    const richSymbols = extractSymbols(tree);

    const symbols: ExtractedSymbol[] = richSymbols.map((symbol, idx) => ({
      nodeId: `${filePath}:${symbol.name}:${idx}`,
      kind: symbol.kind,
      name: symbol.name,
      exported: symbol.exported,
      range: symbol.range,
      signature: symbol.signature,
      visibility: symbol.visibility,
    }));

    return symbols;
  }

  extractImports(
    tree: Tree,
    _content: string,
    _filePath: string,
  ): ExtractedImport[] {
    return extractImports(tree);
  }

  extractCalls(
    tree: Tree,
    _content: string,
    _filePath: string,
    extractedSymbols: ExtractedSymbol[],
  ): ExtractedCall[] {
    return extractCalls(tree, extractedSymbols);
  }
}

function extractSymbols(tree: Parser.Tree): Array<{
  name: string;
  kind: ExtractedSymbol["kind"];
  exported: boolean;
  range: ExtractedSymbol["range"];
  signature?: ExtractedSymbol["signature"];
  visibility?: ExtractedSymbol["visibility"];
}> {
  const symbols: Array<{
    name: string;
    kind: ExtractedSymbol["kind"];
    exported: boolean;
    range: ExtractedSymbol["range"];
    signature?: ExtractedSymbol["signature"];
    visibility?: ExtractedSymbol["visibility"];
  }> = [];

  let currentNamespace: string | null = null;

  function traverse(node: Parser.SyntaxNode): void {
    switch (node.type) {
      case "namespace_definition": {
        const namespaceNode = node.childForFieldName("name");
        if (namespaceNode) {
          currentNamespace = namespaceNode.text;
        }
        break;
      }

      case "class_declaration": {
        const name = extractClassName(node);
        if (name) {
          const fqName = currentNamespace
            ? `${currentNamespace}\\${name}`
            : name;
          const visibility = name.startsWith("_") ? "private" : "public";
          symbols.push({
            name: fqName,
            kind: "class",
            exported: true,
            range: extractRange(node),
            signature: {
              params: [],
            },
            visibility,
          });
        }
        break;
      }

      case "interface_declaration": {
        const name = extractInterfaceName(node);
        if (name) {
          const fqName = currentNamespace
            ? `${currentNamespace}\\${name}`
            : name;
          const visibility = name.startsWith("_") ? "private" : "public";
          symbols.push({
            name: fqName,
            kind: "interface",
            exported: true,
            range: extractRange(node),
            signature: {
              params: [],
            },
            visibility,
          });
        }
        break;
      }

      case "trait_declaration": {
        const name = extractTraitName(node);
        if (name) {
          const fqName = currentNamespace
            ? `${currentNamespace}\\${name}`
            : name;
          const visibility = name.startsWith("_") ? "private" : "public";
          symbols.push({
            name: fqName,
            kind: "class",
            exported: true,
            range: extractRange(node),
            signature: {
              params: [],
            },
            visibility,
          });
        }
        break;
      }

      case "method_declaration": {
        const name = extractMethodName(node);
        if (name) {
          const params = extractMethodParameters(node);
          const returnType = extractMethodReturnType(node);
          const visibility = extractMethodVisibility(node) || "public";

          symbols.push({
            name,
            kind: "method",
            exported: visibility !== "private",
            range: extractRange(node),
            signature: {
              params,
              returns: returnType,
            },
            visibility,
          });
        }
        break;
      }

      case "function_definition": {
        const name = extractFunctionName(node);
        if (name) {
          const fqName = currentNamespace
            ? `${currentNamespace}\\${name}`
            : name;
          const params = extractFunctionParameters(node);
          const returnType = extractFunctionReturnType(node);
          const visibility = name.startsWith("_") ? "private" : "public";

          symbols.push({
            name: fqName,
            kind: "function",
            exported: !name.startsWith("_"),
            range: extractRange(node),
            signature: {
              params,
              returns: returnType,
            },
            visibility,
          });
        }
        break;
      }

      case "property_declaration": {
        const properties = extractPropertyNames(node);
        const visibility = extractPropertyVisibility(node);

        for (const prop of properties) {
          const propVisibility = prop.startsWith("_")
            ? "private"
            : visibility || "public";
          symbols.push({
            name: prop,
            kind: "variable",
            exported: propVisibility !== "private",
            range: extractRange(node),
            visibility: propVisibility,
          });
        }
        break;
      }

      case "const_declaration": {
        const constants = extractConstantNames(node);
        const visibility = extractConstantVisibility(node);

        for (const constant of constants) {
          const constVisibility = visibility || "public";
          symbols.push({
            name: constant,
            kind: "variable",
            exported: constVisibility !== "private",
            range: extractRange(node),
            visibility: constVisibility,
          });
        }
        break;
      }
    }

    for (const child of node.children) {
      traverse(child);
    }
  }

  traverse(tree.rootNode);
  return symbols;
}

function extractClassName(node: Parser.SyntaxNode): string | null {
  const nameNode = node.childForFieldName("name");
  return nameNode?.text || null;
}

function extractInterfaceName(node: Parser.SyntaxNode): string | null {
  const nameNode = node.childForFieldName("name");
  return nameNode?.text || null;
}

function extractTraitName(node: Parser.SyntaxNode): string | null {
  const nameNode = node.childForFieldName("name");
  return nameNode?.text || null;
}

function extractMethodName(node: Parser.SyntaxNode): string | null {
  const nameNode = node.childForFieldName("name");
  return nameNode?.text || null;
}

function extractFunctionName(node: Parser.SyntaxNode): string | null {
  const nameNode = node.childForFieldName("name");
  return nameNode?.text || null;
}

function extractMethodParameters(
  node: Parser.SyntaxNode,
): Array<{ name: string; type?: string }> {
  const params: Array<{ name: string; type?: string }> = [];
  const parametersNode = node.childForFieldName("formal_parameters");

  if (!parametersNode) return params;

  for (const child of parametersNode.children) {
    if (child.type === "simple_parameter") {
      const paramName = extractVariableName(child);
      const paramType = extractParameterType(child);

      if (paramName) {
        params.push({
          name: paramName,
          type: paramType,
        });
      }
    }
  }

  return params;
}

function extractFunctionParameters(
  node: Parser.SyntaxNode,
): Array<{ name: string; type?: string }> {
  const params: Array<{ name: string; type?: string }> = [];
  const parametersNode = node.childForFieldName("formal_parameters");

  if (!parametersNode) return params;

  for (const child of parametersNode.children) {
    if (child.type === "simple_parameter") {
      const paramName = extractVariableName(child);
      const paramType = extractParameterType(child);

      if (paramName) {
        params.push({
          name: paramName,
          type: paramType,
        });
      }
    }
  }

  return params;
}

function extractVariableName(node: Parser.SyntaxNode): string | null {
  const varNode = node.childForFieldName("name");
  if (!varNode || varNode.type !== "variable_name") {
    return null;
  }

  const nameNode = varNode.children.find((c) => c.type === "name");
  return nameNode?.text || null;
}

function extractParameterType(node: Parser.SyntaxNode): string | undefined {
  const typeNode = node.childForFieldName("type");

  if (typeNode) {
    return typeNode.text;
  }

  const child = node.children.find(
    (c) =>
      c.type === "primitive_type" ||
      c.type === "named_type" ||
      c.type === "union_type" ||
      c.type === "optional_type",
  );

  return child?.text || undefined;
}

function extractMethodReturnType(node: Parser.SyntaxNode): string | undefined {
  const returnTypeNode = node.childForFieldName("return_type");
  if (returnTypeNode) {
    return returnTypeNode.text;
  }
  return undefined;
}

function extractFunctionReturnType(
  node: Parser.SyntaxNode,
): string | undefined {
  const returnTypeNode = node.childForFieldName("return_type");
  if (returnTypeNode) {
    return returnTypeNode.text;
  }
  return undefined;
}

function extractMethodVisibility(
  node: Parser.SyntaxNode,
): ExtractedSymbol["visibility"] | null {
  const visibilityNode = node.children.find(
    (c) => c.type === "visibility_modifier",
  );

  if (!visibilityNode) return null;

  const visibility = visibilityNode.text;

  if (
    visibility === "private" ||
    visibility === "protected" ||
    visibility === "public"
  ) {
    return visibility;
  }

  return "public";
}

function extractPropertyVisibility(
  node: Parser.SyntaxNode,
): ExtractedSymbol["visibility"] | null {
  const visibilityNode = node.children.find(
    (c) => c.type === "visibility_modifier",
  );

  if (!visibilityNode) return null;

  const visibility = visibilityNode.text;

  if (
    visibility === "private" ||
    visibility === "protected" ||
    visibility === "public"
  ) {
    return visibility;
  }

  return "public";
}

function extractConstantVisibility(
  node: Parser.SyntaxNode,
): ExtractedSymbol["visibility"] | null {
  const visibilityNode = node.children.find(
    (c) => c.type === "visibility_modifier",
  );

  if (!visibilityNode) return null;

  const visibility = visibilityNode.text;

  if (
    visibility === "private" ||
    visibility === "protected" ||
    visibility === "public"
  ) {
    return visibility;
  }

  return "public";
}

function extractPropertyNames(node: Parser.SyntaxNode): string[] {
  const names: string[] = [];

  for (const child of node.children) {
    if (child.type === "property_element") {
      const varNode = child.children.find((c) => c.type === "variable_name");
      if (varNode && varNode.children.length >= 2) {
        const nameNode = varNode.children.find((c) => c.type === "name");
        if (nameNode) {
          names.push(nameNode.text);
        }
      }
    }
  }

  return names;
}

function extractConstantNames(node: Parser.SyntaxNode): string[] {
  const names: string[] = [];

  for (const child of node.children) {
    if (child.type === "const_element") {
      const nameNode = child.children.find((c) => c.type === "name");
      if (nameNode) {
        names.push(nameNode.text);
      }
    }
  }

  return names;
}

function extractRange(node: Parser.SyntaxNode): ExtractedSymbol["range"] {
  const start = node.startPosition;
  const end = node.endPosition;

  return {
    startLine: start.row + 1,
    startCol: start.column,
    endLine: end.row + 1,
    endCol: end.column,
  };
}

function extractImports(tree: Parser.Tree): ExtractedImport[] {
  const imports: ExtractedImport[] = [];

  const importQuery = createQuery(
    "php",
    `
    (namespace_use_declaration
      (namespace_use_clause
        (qualified_name) @import_name))

    (namespace_use_declaration
      (namespace_use_clause
        (name) @import_name))

    (include_expression
      (string) @include_path)

    (include_once_expression
      (string) @include_path)

    (require_expression
      (string) @include_path)

    (require_once_expression
      (string) @include_path)
  `,
  );

  if (!importQuery) {
    return [];
  }

  const matches = importQuery.matches(tree.rootNode);

  for (const match of matches) {
    for (const capture of match.captures) {
      if (capture.name === "import_name") {
        const node = capture.node;
        const useClause = node.parent;

        if (!useClause || useClause.type !== "namespace_use_clause") {
          continue;
        }

        const statement = useClause.parent;
        if (!statement || statement.type !== "namespace_use_declaration") {
          continue;
        }

        const specifier = node.text;
        const isRelative = specifier.startsWith("\\");
        const isExternal = true;

        const result: ExtractedImport = {
          specifier,
          isRelative,
          isExternal,
          imports: [],
          isReExport: false,
        };

        const aliasingClause = useClause.children.find(
          (c) => c.type === "namespace_aliasing_clause",
        );

        if (aliasingClause) {
          const aliasNameNode = aliasingClause.children.find(
            (c) => c.type === "name",
          );
          if (aliasNameNode) {
            result.imports.push(aliasNameNode.text);
          }
        } else {
          const parts = specifier.split("\\");
          result.imports.push(parts[parts.length - 1]);
        }

        imports.push(result);
      } else if (capture.name === "include_path") {
        const node = capture.node;
        const statement = node.parent;

        if (!statement) {
          continue;
        }

        const specifier = node.text.replace(/^["']|["']$/g, "");
        const isRelative =
          specifier.startsWith("./") || specifier.startsWith("../");
        const isExternal = !isRelative;

        const result: ExtractedImport = {
          specifier,
          isRelative,
          isExternal,
          imports: [],
          isReExport: false,
        };

        imports.push(result);
      }
    }
  }

  return imports;
}

function extractCalls(
  tree: Parser.Tree,
  extractedSymbols: ExtractedSymbol[],
): ExtractedCall[] {
  const calls: ExtractedCall[] = [];
  const seenCallNodes = new Set<number>();

  const symbolMap = new Map<string, ExtractedSymbol>();
  for (const symbol of extractedSymbols) {
    symbolMap.set(symbol.name, symbol);
  }

  const callQuery = createQuery(
    "php",
    `(function_call_expression
      (name) @callee)

    (function_call_expression
      (variable_name) @var_callee)

    (member_call_expression
      (name) @attr
      (variable_name) @recv)

    (member_call_expression
      (name) @var_attr
      (variable_name) @recv)

    (scoped_call_expression
      (name) @scope
      (name) @static_callee)

    (scoped_call_expression
      (qualified_name) @qualified_scope
      (name) @static_callee)
  `,
  );

  if (!callQuery) {
    return [];
  }

  const matches = callQuery.matches(tree.rootNode);

  for (const match of matches) {
    const callNode = match.captures[0].node.parent;

    if (!callNode) continue;

    const nodeId = callNode.id;
    if (seenCallNodes.has(nodeId)) continue;
    seenCallNodes.add(nodeId);

    const calleeCapture = match.captures.find((c) => c.name === "callee");
    const varCalleeCapture = match.captures.find(
      (c) => c.name === "var_callee",
    );
    const attrCapture = match.captures.find((c) => c.name === "attr");
    const varAttrCapture = match.captures.find((c) => c.name === "var_attr");
    const recvCapture = match.captures.find((c) => c.name === "recv");
    const scopeCapture = match.captures.find((c) => c.name === "scope");
    const qualifiedScopeCapture = match.captures.find(
      (c) => c.name === "qualified_scope",
    );
    const staticCalleeCapture = match.captures.find(
      (c) => c.name === "static_callee",
    );

    let calleeIdentifier = "";
    let callType: ExtractedCall["callType"] = "function";
    let isResolved = false;
    let calleeSymbolId: string | undefined;

    if (calleeCapture) {
      calleeIdentifier = calleeCapture.node.text;
      const symbol = symbolMap.get(calleeIdentifier);
      if (symbol && (symbol.kind === "function" || symbol.kind === "method")) {
        isResolved = true;
        calleeSymbolId = symbol.nodeId;
      }
    } else if (varCalleeCapture) {
      calleeIdentifier = varCalleeCapture.node.text;
      callType = "dynamic";
      isResolved = false;
    } else if (attrCapture && recvCapture) {
      calleeIdentifier = `${recvCapture.node.text}.${attrCapture.node.text}`;
      callType = "method";

      if (recvCapture.node.text === "$this") {
        const symbol = symbolMap.get(attrCapture.node.text);
        if (symbol && symbol.kind === "method") {
          isResolved = true;
          calleeSymbolId = symbol.nodeId;
        }
      }
    } else if (varAttrCapture && recvCapture) {
      calleeIdentifier = `${recvCapture.node.text}.${varAttrCapture.node.text}`;
      callType = "dynamic";
      isResolved = false;
    } else if ((scopeCapture || qualifiedScopeCapture) && staticCalleeCapture) {
      const scopeText = scopeCapture
        ? scopeCapture.node.text
        : qualifiedScopeCapture!.node.text;
      calleeIdentifier = `${scopeText}::${staticCalleeCapture.node.text}`;
      callType = "function";

      const fullMethodName = `${scopeText}::${staticCalleeCapture.node.text}`;
      const symbol = symbolMap.get(fullMethodName);
      if (symbol && symbol.kind === "method") {
        isResolved = true;
        calleeSymbolId = symbol.nodeId;
      }
    }

    const callerNodeId = findEnclosingSymbolUtil(callNode, extractedSymbols);

    calls.push({
      callerNodeId,
      calleeIdentifier,
      isResolved,
      callType,
      calleeSymbolId,
      range: extractRange(callNode),
    });
  }

  return calls;
}

function clearCache(): void {
  clearGrammarCache("php");
}

export { PhpAdapter, clearCache };

File: native/src/extract/symbols/php.rs
use tree_sitter::Node;

use crate::types::NativeParsedSymbol;

use super::common::{find_child_by_kind, find_child_node, make_symbol, node_text, ParamInfo};

pub fn extract_symbols_php(
    root: Node<'_>,
    source: &[u8],
    repo_id: &str,
    rel_path: &str,
) -> Vec<NativeParsedSymbol> {
    let mut symbols = Vec::new();
    let mut stack = vec![root];
    let mut current_namespace: Option<String> = None;

    while let Some(node) = stack.pop() {
        match node.kind() {
            "namespace_definition" => {
                process_namespace_definition(
                    node,
                    source,
                    repo_id,
                    rel_path,
                    &mut current_namespace,
                    &mut symbols,
                );
            }
            "class_declaration" => {
                if let Some(symbol) = process_type_like(
                    node,
                    source,
                    repo_id,
                    rel_path,
                    "class",
                    current_namespace.as_deref(),
                    false,
                ) {
                    symbols.push(symbol);
                }
            }
            "interface_declaration" => {
                if let Some(symbol) = process_type_like(
                    node,
                    source,
                    repo_id,
                    rel_path,
                    "interface",
                    current_namespace.as_deref(),
                    false,
                ) {
                    symbols.push(symbol);
                }
            }
            "trait_declaration" => {
                if let Some(symbol) = process_type_like(
                    node,
                    source,
                    repo_id,
                    rel_path,
                    "class",
                    current_namespace.as_deref(),
                    true,
                ) {
                    symbols.push(symbol);
                }
            }
            "method_declaration" => {
                if let Some(symbol) = process_method_declaration(node, source, repo_id, rel_path) {
                    symbols.push(symbol);
                }
            }
            "function_definition" => {
                if let Some(symbol) = process_function_definition(
                    node,
                    source,
                    repo_id,
                    rel_path,
                    current_namespace.as_deref(),
                ) {
                    symbols.push(symbol);
                }
            }
            "property_declaration" => {
                process_property_declaration(node, source, repo_id, rel_path, &mut symbols);
            }
            "const_declaration" => {
                process_const_declaration(node, source, repo_id, rel_path, &mut symbols);
            }
            _ => {}
        }

        let child_count = node.child_count();
        for i in (0..child_count).rev() {
            if let Some(child) = node.child(i) {
                stack.push(child);
            }
        }
    }

    symbols
}

fn process_namespace_definition(
    node: Node<'_>,
    source: &[u8],
    repo_id: &str,
    rel_path: &str,
    current_namespace: &mut Option<String>,
    symbols: &mut Vec<NativeParsedSymbol>,
) {
    let namespace_node = node
        .child_by_field_name("name")
        .or_else(|| find_child_node(node, "namespace_name"))
        .or_else(|| find_child_node(node, "qualified_name"))
        .or_else(|| find_child_node(node, "name"));

    let Some(namespace_node) = namespace_node else {
        return;
    };

    let namespace = node_text(namespace_node, source).trim().to_string();
    if namespace.is_empty() {
        return;
    }

    *current_namespace = Some(namespace.clone());

    let mut symbol = make_symbol(
        &namespace,
        "module",
        node,
        source,
        repo_id,
        rel_path,
        &[],
        None,
        &[],
        "public",
        &[],
    );
    symbol.exported = true;
    symbols.push(symbol);
}

fn process_type_like(
    node: Node<'_>,
    source: &[u8],
    repo_id: &str,
    rel_path: &str,
    kind: &str,
    current_namespace: Option<&str>,
    is_trait: bool,
) -> Option<NativeParsedSymbol> {
    let name = extract_name(node, source)?;
    let fq_name = if let Some(namespace) = current_namespace {
        format!(r"{namespace}\{name}")
    } else {
        name.clone()
    };

    let visibility = if name.starts_with('_') {
        "private".to_string()
    } else {
        "public".to_string()
    };

    let mut symbol = make_symbol(
        &fq_name,
        kind,
        node,
        source,
        repo_id,
        rel_path,
        &[],
        None,
        &[],
        &visibility,
        &[],
    );
    symbol.exported = true;

    if is_trait {
        symbol.role_tags.push("trait".to_string());
    }

    Some(symbol)
}

fn process_method_declaration(
    node: Node<'_>,
    source: &[u8],
    repo_id: &str,
    rel_path: &str,
) -> Option<NativeParsedSymbol> {
    let name = extract_name(node, source)?;
    let params = extract_parameters(node, source);
    let returns = extract_return_type(node, source);
    let visibility = extract_visibility(node, source).unwrap_or_else(|| "public".to_string());

    let mut symbol = make_symbol(
        &name,
        "method",
        node,
        source,
        repo_id,
        rel_path,
        &params,
        returns.as_deref(),
        &[],
        &visibility,
        &[],
    );
    symbol.exported = visibility != "private";
    Some(symbol)
}

fn process_function_definition(
    node: Node<'_>,
    source: &[u8],
    repo_id: &str,
    rel_path: &str,
    current_namespace: Option<&str>,
) -> Option<NativeParsedSymbol> {
    let name = extract_name(node, source)?;
    let fq_name = if let Some(namespace) = current_namespace {
        format!(r"{namespace}\{name}")
    } else {
        name.clone()
    };

    let params = extract_parameters(node, source);
    let returns = extract_return_type(node, source);
    let visibility = if name.starts_with('_') {
        "private".to_string()
    } else {
        "public".to_string()
    };

    let mut symbol = make_symbol(
        &fq_name,
        "function",
        node,
        source,
        repo_id,
        rel_path,
        &params,
        returns.as_deref(),
        &[],
        &visibility,
        &[],
    );
    symbol.exported = !name.starts_with('_');
    Some(symbol)
}

fn process_property_declaration(
    node: Node<'_>,
    source: &[u8],
    repo_id: &str,
    rel_path: &str,
    symbols: &mut Vec<NativeParsedSymbol>,
) {
    let properties = extract_property_names(node, source);
    let visibility = extract_visibility(node, source).unwrap_or_else(|| "public".to_string());

    for property in properties {
        let normalized = property.trim_start_matches('$');
        let property_visibility = if normalized.starts_with('_') {
            "private".to_string()
        } else {
            visibility.clone()
        };

        let mut symbol = make_symbol(
            &property,
            "variable",
            node,
            source,
            repo_id,
            rel_path,
            &[],
            None,
            &[],
            &property_visibility,
            &[],
        );
        symbol.exported = property_visibility != "private";
        symbols.push(symbol);
    }
}

fn process_const_declaration(
    node: Node<'_>,
    source: &[u8],
    repo_id: &str,
    rel_path: &str,
    symbols: &mut Vec<NativeParsedSymbol>,
) {
    let constants = extract_constant_names(node, source);
    let visibility = extract_visibility(node, source).unwrap_or_else(|| "public".to_string());

    for constant in constants {
        let mut symbol = make_symbol(
            &constant,
            "variable",
            node,
            source,
            repo_id,
            rel_path,
            &[],
            None,
            &[],
            &visibility,
            &[],
        );
        symbol.exported = visibility != "private";
        symbols.push(symbol);
    }
}

fn extract_name(node: Node<'_>, source: &[u8]) -> Option<String> {
    let name_node = node
        .child_by_field_name("name")
        .or_else(|| find_child_node(node, "name"));
    let name = name_node
        .map(|n| node_text(n, source).to_string())
        .unwrap_or_default();
    (!name.is_empty()).then_some(name)
}

fn extract_parameters(node: Node<'_>, source: &[u8]) -> Vec<ParamInfo> {
    let parameters_node = node
        .child_by_field_name("formal_parameters")
        .or_else(|| find_child_node(node, "formal_parameters"));
    let Some(parameters_node) = parameters_node else {
        return Vec::new();
    };

    let mut params = Vec::new();
    let mut cursor = parameters_node.walk();
    for child in parameters_node.children(&mut cursor) {
        if child.kind() != "simple_parameter" {
            continue;
        }

        let name = child
            .child_by_field_name("name")
            .and_then(|n| extract_variable_name(n, source));
        let Some(name) = name else {
            continue;
        };

        let type_annotation = extract_parameter_type(child, source);
        params.push(ParamInfo {
            name,
            type_annotation,
        });
    }

    params
}

fn extract_variable_name(node: Node<'_>, source: &[u8]) -> Option<String> {
    if node.kind() == "variable_name" {
        if let Some(inner_name) = find_child_by_kind(node, "name", source) {
            return (!inner_name.is_empty()).then_some(inner_name);
        }
        let raw = node_text(node, source).to_string();
        let trimmed = raw.trim_start_matches('$').to_string();
        return (!trimmed.is_empty()).then_some(trimmed);
    }

    if node.kind() == "name" {
        let text = node_text(node, source).to_string();
        return (!text.is_empty()).then_some(text);
    }

    None
}

fn extract_parameter_type(node: Node<'_>, source: &[u8]) -> Option<String> {
    if let Some(type_node) = node.child_by_field_name("type") {
        let text = node_text(type_node, source).to_string();
        if !text.is_empty() {
            return Some(text);
        }
    }

    let mut cursor = node.walk();
    for child in node.children(&mut cursor) {
        if matches!(
            child.kind(),
            "primitive_type" | "named_type" | "union_type" | "optional_type"
        ) {
            let text = node_text(child, source).to_string();
            if !text.is_empty() {
                return Some(text);
            }
        }
    }

    None
}

fn extract_return_type(node: Node<'_>, source: &[u8]) -> Option<String> {
    let return_type = node
        .child_by_field_name("return_type")
        .or_else(|| find_child_node(node, "return_type"));
    let text = return_type
        .map(|n| node_text(n, source).to_string())
        .unwrap_or_default();
    (!text.is_empty()).then_some(text)
}

fn extract_visibility(node: Node<'_>, source: &[u8]) -> Option<String> {
    let mut cursor = node.walk();
    for child in node.children(&mut cursor) {
        if child.kind() != "visibility_modifier" {
            continue;
        }

        let text = node_text(child, source);
        if text == "public" || text == "private" || text == "protected" {
            return Some(text.to_string());
        }
        return Some("public".to_string());
    }

    None
}

fn extract_property_names(node: Node<'_>, source: &[u8]) -> Vec<String> {
    let mut names = Vec::new();
    let mut cursor = node.walk();

    for child in node.children(&mut cursor) {
        if child.kind() != "property_element" {
            continue;
        }

        let Some(variable_node) = find_child_node(child, "variable_name") else {
            continue;
        };

        let Some(name) = extract_variable_name(variable_node, source) else {
            continue;
        };

        names.push(format!("${name}"));
    }

    names
}

fn extract_constant_names(node: Node<'_>, source: &[u8]) -> Vec<String> {
    let mut names = Vec::new();
    let mut cursor = node.walk();

    for child in node.children(&mut cursor) {
        if child.kind() != "const_element" {
            continue;
        }

        if let Some(name) = find_child_by_kind(child, "name", source) {
            if !name.is_empty() {
                names.push(name);
            }
        }
    }

    names
}
